src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RSABCrypt.cs

```diff
@@ -0,0 +1,427 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Threading;
+using Internal.NativeCrypto;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.Security.Cryptography
+{
+    internal sealed class RSABCrypt : RSA
+    {
+        private static readonly SafeBCryptAlgorithmHandle s_algHandle =
+            Interop.BCrypt.BCryptOpenAlgorithmProvider(BCryptNative.AlgorithmName.RSA);
+
+        // See https://msdn.microsoft.com/en-us/library/windows/desktop/bb931354(v=vs.85).aspx
+        // All values are in bits.
+        private static readonly KeySizes s_keySizes =
+            new KeySizes(minSize: 512, maxSize: 16384, skipSize: 64);
+
+        private SafeBCryptKeyHandle? _key;
+        private int _lastKeySize;
+
+        internal RSABCrypt()
+        {
+            KeySizeValue = 2048;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                _key?.Dispose();
+            }
+
+            _lastKeySize = -1;
+        }
+
+        private SafeBCryptKeyHandle GetKey()
+        {
+            int keySize = KeySize;
+
+            // Since _lastKeySize also tracks the disposal state, we can do the equals check first.
+            if (_lastKeySize == keySize)
+            {
+                Debug.Assert(_key != null);
+                return _key;
+            }
+
+            ThrowIfDisposed();
+
+            SafeBCryptKeyHandle newKey = Interop.BCrypt.BCryptGenerateKeyPair(s_algHandle, keySize);
+            Interop.BCrypt.BCryptFinalizeKeyPair(newKey);
+            SetKey(newKey);
+            return newKey;
+        }
+
+        private void SetKey(SafeBCryptKeyHandle newKey)
+        {
+            Debug.Assert(!newKey.IsInvalid);
+
+            int keySize = Interop.BCrypt.BCryptGetDWordProperty(
+                newKey,
+                Interop.BCrypt.BCryptPropertyStrings.BCRYPT_KEY_STRENGTH);
+
+            SafeBCryptKeyHandle? oldKey = Interlocked.Exchange(ref _key, newKey);
+            ForceSetKeySize(keySize);
+            oldKey?.Dispose();
+        }
+
+        public override RSAParameters ExportParameters(bool includePrivateParameters)
+        {
+            SafeBCryptKeyHandle key = GetKey();
+
+            ArraySegment<byte> keyBlob = Interop.BCrypt.BCryptExportKey(
+                key,
+                includePrivateParameters ?
+                    Interop.BCrypt.KeyBlobType.BCRYPT_RSAFULLPRIVATE_BLOB :
+                    Interop.BCrypt.KeyBlobType.BCRYPT_RSAPUBLIC_KEY_BLOB);
+
+            RSAParameters ret = default;
+            ret.FromBCryptBlob(keyBlob, includePrivateParameters);
+            CryptoPool.Return(keyBlob);
```

Comment:
Should this `Return` be in a finally since the blob may contain private parameters and we want to zero them? (Like `ImportParameters` below)

Reply:
Hm. My logic behind having them different is that ImportParameters comes from user-supplied data.  Users are notoriously bad at providing accurate data, so there's a fair chance that ToBCryptBlob will succeed, but BCryptImportKeyPair will fail.  We know that the array isn't still in use, so it's safe to return to the pool, so let's do so (also, bonus: clearing out copies of D/P/Q we may have made).

In ExportParameters we believe we handle the output forms of BCRYPT_RSAFULLPRIVATE_BLOB and BCRYPT_RSAPUBLIC_KEY_BLOB.  The only thing that can really go wrong is something like OutOfMemoryException... so the try/finally isn't justified.